dali/operators/reader/loader/filesystem.h

```diff
@@ -23,16 +24,32 @@
 namespace dali {
 namespace filesystem {
 
-DLL_PUBLIC vector<string> traverse_directories(const string &path, const string &filter);
+inline bool starts_with(const std::string &str, const char *prefix) {
+  // TODO(janton): this is a substitute for C++20's string::starts_with
+  // trick: this only matches if the prefix is found at the beginning of the string
+  return str.rfind(prefix, 0) == 0;
+}
+
+struct FileLabelEntry {
+  std::string filename;
+  // only if read_label==true
+  std::optional<int> label = {};
+  // only populated when size is known without opening (e.g. s3)
+  std::optional<std::size_t> size = {};
+};
+
+struct TraverseDirectoriesOptions {
+  bool read_label = true;                 // expect each subdirectory to have one label
+  bool case_sensitive_filter = false;     // whether the filter patterns are case-sensitive
+  std::vector<std::string> file_filters;  // pattern to apply to filenames
+  std::vector<std::string> dir_filters;   // pattern to apply to subdirectories
+};
 
 /**
- * @brief Finds all (file, label) pairs matching any filter from the list.
+ * @brief Finds all (file, label, size) information, following the criteria given by opts.
  */
-DLL_PUBLIC vector<std::pair<string, int>> traverse_directories(
-    const string &file_root,
-    const vector<string> &filters,
-    bool case_sensitive_filter = false,
-    const vector<string>& dir_filters = {});
+DLL_PUBLIC vector<FileLabelEntry> traverse_directories(const std::string &file_root,
+                                                       const TraverseDirectoriesOptions &opts);
```

Comment:
```suggestion
DLL_PUBLIC vector<FileLabelEntry> discover_files(const std::string &file_root,
                                                       const TraverseDirectoriesOptions &opts);
```
or even
```suggestion
DLL_PUBLIC vector<FileLabelEntry> discover_dataset(const std::string &file_root,
                                                       const TraverseDirectoriesOptions &opts);
```
...because we're discovering labels alongside the files.